### PR TITLE
Read iKey from environment variable (APPINSIGHTS_INSTRUMENTATIONKEY)

### PR DESCRIPTION
**Pull request by abaranch**
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/136

### CODE REVIEW

File: src/Core/Managed/PCL/Extensibility/Implementation/Platform/PlatformImplementation.cs [hunk 1]
```diff
@@ -35,5 +35,10 @@ public IDebugOutput GetDebugOutput()
             
             return this.debugOutput;
         }
+
+        public string GetEnvironmentVariable(string name)
+        {
+            return null;
```
Comment (dnduffy) [inline src/Core/Managed/PCL/Extensibility/Implementation/Platform/PlatformImplementation.cs]: Is there maybe a .NET Core equivalent for getting environment variables we could use?

File: src/Core/Managed/Shared/TelemetryClient.cs [hunk 2]
```diff
@@ -371,6 +373,11 @@ public void Initialize(ITelemetry telemetry)
             if (string.IsNullOrEmpty(instrumentationKey))
             {
                 instrumentationKey = this.configuration.InstrumentationKey;
+
+                if (string.IsNullOrEmpty(instrumentationKey))
+                {
+                    instrumentationKey = PlatformSingleton.Current.GetEnvironmentVariable(InstrumentationKeyWebSitesEnvironmentVariable);
```
Comment (SergeyKanzhelev) [inline src/Core/Managed/Shared/TelemetryClient.cs]: So this ikey has the least priority? Is it what we want for Web Apps case? In Web Apps you can configure ikey in `AI.config` file and then override it using Web App settings. Not other way around.

Also I'm concerned about performance of it. Maybe just read it once and cache? Or we need to support the scenario when environment variable will be changed later?

File: src/Core/Managed/PCL/Extensibility/Implementation/Platform/PlatformImplementation.cs (hunk 1 above)
Comment (dnduffy) [inline src/Core/Managed/PCL/Extensibility/Implementation/Platform/PlatformImplementation.cs]: Maybe something in Microsoft.Extensions.Configuration.EnvironmentVariables or Microsoft.Extensions.Configuration could be used.

File: src/Core/Managed/Shared/TelemetryClient.cs (hunk 2 above)
Comment (dnduffy) [inline src/Core/Managed/Shared/TelemetryClient.cs]: I agree on caching, discussing the priority ordering, but I don't think it should change during runtime.

File: src/Core/Managed/PCL/Extensibility/Implementation/Platform/PlatformImplementation.cs (hunk 1 above)
Comment (abaranch) [inline src/Core/Managed/PCL/Extensibility/Implementation/Platform/PlatformImplementation.cs]: Not in PCL. Those are .Net Core specific namespaces. If ever change target framework we can start supporting it.

Comment (dnduffy) [inline src/Core/Managed/PCL/Extensibility/Implementation/Platform/PlatformImplementation.cs]: Good point, I keep thinking of .NET Core as a replacement for PCL but it is completely new and isn't portable backwards (as such).

File: src/Core/Managed/Shared/TelemetryClient.cs (hunk 2 above)
Comment (abaranch) [inline src/Core/Managed/Shared/TelemetryClient.cs]: Yes, priority order is correct for web sites (@iusafaro)
Resolved caching.

Comment (SergeyKanzhelev) [inline src/Core/Managed/Shared/TelemetryClient.cs]: @iusafaro, this priority order doesn't make sense to me. Can you pls elaborate why it is done this way?

Comment (abaranch) [inline src/Core/Managed/Shared/TelemetryClient.cs]: Per offline discussion we will change the order so it is: code -> environment variable -> config

**Comment by SergeyKanzhelev** [comment]
@iusafaro can you please ensure that we do not need to re-read instrumentation key from the environment variable after Web App swap?

File: src/Core/Managed/Shared/TelemetryClient.cs [hunk 3]
```diff
@@ -370,7 +372,12 @@ public void Initialize(ITelemetry telemetry)
 
             if (string.IsNullOrEmpty(instrumentationKey))
             {
-                instrumentationKey = this.configuration.InstrumentationKey;
+                instrumentationKey = PlatformSingleton.Current.GetEnvironmentVariable(InstrumentationKeyWebSitesEnvironmentVariable);
+                
+                if (string.IsNullOrEmpty(instrumentationKey))
+                {
+                    instrumentationKey = this.configuration.InstrumentationKey;
```
Comment (SergeyKanzhelev) [inline src/Core/Managed/Shared/TelemetryClient.cs]: So we doing it before initializers have a chance to override it?

Comment (abaranch) [inline src/Core/Managed/Shared/TelemetryClient.cs]: Yes

Comment (SergeyKanzhelev) [inline src/Core/Managed/Shared/TelemetryClient.cs]: Thinking about it - I guess it is fine

**Pull request merged**